### PR TITLE
[Merged by Bors] - feat(Data/Finsupp): add notation

### PR DESCRIPTION
**Pull request by eric-wieser**
This PR provides `fun₀ | 3 => a | 7 => b` notation for `Finsupp`, which desugars to `Finsupp.update` and `Finsupp.single`, in the same way that `{a, b}` desugars to `insert` and `singleton`.

It also provides a matching delaborator: as an example of the effect of this, `Finsupp.single_add` now shows as:
```lean
Finsupp.single_add.{u_2, u_1} {α : Type u_1} {M : Type u_2} [inst✝ : AddZeroClass M] (a : α) (b₁ b₂ : M) :
  (fun₀ | a => b₁ + b₂) = (fun₀ | a => b₁) + fun₀ | a => b₂
```

Finally, it provides a `Repr` instance; though this is somewhat misleading as the syntax that is produced by `Repr` isn't actually computable so can't round-trip!

[Discussed on Zulip here](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Notation.20for.20.60Finsupp.60/near/381919585).

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

A nice example of the `Repr` in action is
```lean
#eval Finsupp.antidiagonal ⟨{1, 2}, fun x : Fin 3 => x, sorry⟩
/-
{(0, fun₀ | 1 => 1 | 2 => 2),
 (fun₀ | 1 => 1, fun₀ | 2 => 2),
 (fun₀ | 2 => 1, fun₀ | 1 => 1 | 2 => 1),
 (fun₀ | 1 => 1 | 2 => 1, fun₀ | 2 => 1),
 (fun₀ | 2 => 2, fun₀ | 1 => 1),
 (fun₀ | 1 => 1 | 2 => 2, 0)}
-/
```

### CODE REVIEW

**Comment by thorimur** [comment]
I really like this! :) Given that `↦` is standard for `fun` expressions in mathlib, though, I would suggest that:
1. both `=>` and `↦` should be available as input notation, and
2. the delaborator should use `↦`
so that it's as close as possible to `fun` syntax.

**Comment by eric-wieser** [comment]
> 1. both `=>` and `↦` should be available as input notation, and

This does not work for regular functions, so I think supporting it here would just be confusing
```
#check show ℕ → ℕ from fun | 0 => 0 | Nat.succ n => n + 2 -- ok
#check show ℕ → ℕ from fun | 0 ↦ 0 | Nat.succ n ↦ n + 2 -- fail
```
If we fix `fun` then `fun₀` will start working automatically

> 2. the delaborator and `Repr` instance should use `↦`

Not without 1. they shouldn't!

File: Mathlib/Data/Finsupp/Notation.lean [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+/-- Unexpander for the `fun₀ | i => x` notation. -/
+@[app_unexpander Finsupp.update]
+def updateUnexpander : Lean.PrettyPrinter.Unexpander
+  | `($_ $f $pat $val) => match f with
+    | `(fun₀ $xs:matchAlt*) => `(fun₀ $xs:matchAlt* | $pat => $val)
+    | _ => throw ()
+  | _ => throw ()
+
+/-- Display `Finsupp` using `fun₀` notation. -/
+unsafe instance {α β} [Repr α] [Repr β] [Zero β] : Repr (α →₀ β) where
```
Comment (semorrison) [inline Mathlib/Data/Finsupp/Notation.lean:61]: Better to use `Std.Util.TermUnsafe` here, rather than propagating the `unsafe`, I think.

Comment (eric-wieser) [reply Mathlib/Data/Finsupp/Notation.lean]: Done

Comment (eric-wieser) [reply Mathlib/Data/Finsupp/Notation.lean]: Couldn't you argue that `Quot.unquot` should be made into a `safe` wrapper for the same reason?

Comment (digama0) [reply Mathlib/Data/Finsupp/Notation.lean:61]: Oh, BTW there is another trick you can use here: although alpha might not be ordered, `Format` is, so you could just format them first and then sort the results to make the output order independent.

Comment (eric-wieser) [reply Mathlib/Data/Finsupp/Notation.lean:61]: I decided not to do that for the same reason that the multiset repr doesn't; it gives stupid `{1, 10, 11, 2, 3}` orderings. I have an open issue about doing a better job of this somewhere

**Review by kmill — COMMENTED** [review]
I think it would be nicer to use `syntax` commands instead of using `leading_parser` directly, but we don't need to change it for trying it out in mathlib.

Note that the documentation you've put on `fun₀` is just ` Finsupp`'s documentation. At some point we should have documentation on the notation that explains the notation itself.

bors r+

**Pull request closed without merging**